### PR TITLE
Adjoint typehints

### PR DESCRIPTION
**Pull request by smartalecH**
This PR focuses on type hints for the adjoint model. Future PRs should focus on the
* Docstrings 
* Adjoint overview page in docs

### CODE REVIEW

File: python/adjoint/utils.py [hunk 1]
```diff
@@ -22,6 +22,11 @@
 # default finite difference step size when calculating Aᵤ
 FD_DEFAULT = 1e-3
 
+# Type hint for all components
+MEEP_COMPONENTS = Union[
+    mp.Ex, mp.Ey, mp.Ez, mp.Hx, mp.Hy, mp.Hz, mp.Dx, mp.Dy, mp.Dz, mp.Bx, mp.By, mp.Bz
```
Comment (ianwilliamson) [inline python/adjoint/utils.py:27]: FYI- These are instances of integers, not types. I think the actual Python type for a field component in Meep is an int.

Comment (smartalecH) [reply python/adjoint/utils.py]: Fixed, thanks @ianwilliamson!

File: python/adjoint/optimization_problem.py [hunk 2]
```diff
@@ -24,21 +25,28 @@ class OptimizationProblem:
...
-        nf=None,
-        decay_by=1e-11,
-        decimation_factor=0,
-        minimum_run_time=0,
-        maximum_run_time=None,
-        finite_difference_step=utils.FD_DEFAULT,
+        simulation: mp.Simulation,
+        objective_functions: List[Callable],
+        objective_arguments: List[ObjectiveQuantity],
+        design_regions: [DesignRegion],
```
Comment (oskooi) [inline python/adjoint/optimization_problem.py:31]: Should this be `List[DesignRegion]`?

File: python/adjoint/optimization_problem.py [hunk 3]
```diff
@@ -24,21 +25,28 @@ class OptimizationProblem:
...
+        decay_by: Optional[float] = 1e-11,
+        decimation_factor: Optional[int] = 0,
+        minimum_run_time: Optional[float] = 0,
+        maximum_run_time: Optional[float] = None,
+        finite_difference_step: Optional[float] = utils.FD_DEFAULT,
         step_funcs: list = None,
     ):
+        """
+        + **`simulation` [ `Simulation` ]** — The corresponding Meep
+        `Simulation` object that describeds the problem (e.g. sources,
```
Comment (oskooi) [inline python/adjoint/optimization_problem.py:45]: Change `describeds` to `describes`.

File: python/adjoint/utils.py [hunk 4]
```diff
@@ -1,4 +1,4 @@
-from typing import Iterable, List, Tuple
+from typing import Iterable, List, Tuple, Union
```
Comment (oskooi) [inline python/adjoint/utils.py:1]: Why is `Union` necessary?

File: python/adjoint/objective.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
 """Handling of objective functions and objective quantities."""
 import abc
 from collections import namedtuple
+from typing import Callable, List, Union, Optional
```
Comment (oskooi) [inline python/adjoint/objective.py:4]: I don't think `Union` is necessary.

File: python/adjoint/optimization_problem.py [hunk 6]
```diff
@@ -1,11 +1,12 @@
 from collections import namedtuple
+from typing import Callable, List, Union, Optional, Tuple
```
Comment (oskooi) [inline python/adjoint/optimization_problem.py:2]: Remove `Union` and `Tuple`?

File: python/adjoint/optimization_problem.py (hunk 2 above)
Comment (smartalecH) [reply python/adjoint/optimization_problem.py]: Fixed.

File: python/adjoint/optimization_problem.py (hunk 3 above)
Comment (smartalecH) [reply python/adjoint/optimization_problem.py]: Fixed.

File: python/adjoint/utils.py (hunk 4 above)
Comment (smartalecH) [reply python/adjoint/utils.py]: Fixed.

File: python/adjoint/objective.py (hunk 5 above)
Comment (smartalecH) [reply python/adjoint/objective.py]: Fixed.

File: python/adjoint/optimization_problem.py (hunk 6 above)
Comment (smartalecH) [reply python/adjoint/optimization_problem.py:2]: Both are used here.

**Review by oskooi — APPROVED** [review]

**Pull request merged**